solana/pyth2wormhole/client/src/main.rs

```diff
@@ -225,181 +245,307 @@ async fn main() -> Result<(), ErrBox> {
+            symbols: batch_state.symbols.clone(),
             n_retries,
-            retry_interval,
-            daemon,
-            rpc_cfg.clone(),
-            p2w_addr,
-            config.clone(),
-            Keypair::from_bytes(&payer.to_bytes()).unwrap(),
-            message_q_mtx.clone(),
-        )
+            retry_interval: retry_interval.clone(),
+            rpc_cfg: rpc_cfg.clone(),
+            p2w_addr: p2w_addr.clone(),
+            p2w_config: p2w_cfg.clone(),
+            payer: Keypair::from_bytes(&payer.to_bytes()).unwrap(),
+            message_q_mtx: message_q_mtx.clone(),
+        })
     });
 
-    info!("Spinning up attestation sched jobs");
-
-    let results = futures::future::join_all(attestation_sched_futs).await; // May never finish for daemon mode
+    let results = futures::future::join_all(retry_jobs).await;
 
-    info!("Got {} results", results.len());
-
-    // With daemon mode off, the sched jobs return from the
-    // join_all. We filter out errors and report them
+    // After completing, we count any errors coming from the sched
+    // futs.
     let errors: Vec<_> = results
         .iter()
         .enumerate()
         .filter_map(|(idx, r)| {
             r.as_ref()
                 .err()
-                .map(|e| format!("Error {}: {:#?}\n", idx + 1, e))
+                .map(|e| format!("Error {}: {:?}\n", idx + 1, e))
         })
         .collect();
 
     if !errors.is_empty() {
         let err_lines = errors.join("\n");
-        let msg = format!(
-            "{} of {} batches failed:\n{}",
-            errors.len(),
-            batch_count,
-            err_lines
-        );
+        let msg = format!("{} batches failed:\n{}", errors.len(), err_lines);
         error!("{}", msg);
         return Err(msg.into());
     }
-
     Ok(())
 }
 
-#[derive(Clone)]
-pub struct RpcCfg {
-    pub url: String,
-    pub timeout: Duration,
-    pub commitment: CommitmentConfig,
+/// Constructs attestation scheduling jobs from attestation config.
+fn prepare_attestation_sched_jobs(
+    attestation_cfg: &AttestationConfig,
+    p2w_cfg: &Pyth2WormholeConfig,
+    rpc_cfg: &Arc<RLMutex<RpcCfg>>,
+    p2w_addr: &Pubkey,
+    payer: &Keypair,
+) -> futures::future::JoinAll<impl Future<Output = Result<(), ErrBoxSend>>> {
+    info!(
+        "{} symbol groups read, dividing into batches",
+        attestation_cfg.symbol_groups.len(),
+    );
+
+    // Flatten attestation config into a plain list of batches
+    let batches: Vec<_> = attestation_cfg.as_batches(p2w_cfg.max_batch_size as usize);
+
+    let batch_count = batches.len();
+
+    // For enforcing min_msg_reuse_interval_ms, we keep a piece of
+    // state that creates or reuses accounts if enough time had
+    // passed
+    let message_q_mtx = Arc::new(Mutex::new(P2WMessageQueue::new(
+        Duration::from_millis(attestation_cfg.min_msg_reuse_interval_ms),
+        attestation_cfg.max_msg_accounts as usize,
+    )));
```

Comment:
I think the state of this queue gets reset each time the attestation batches change. Does that mean we can end up reusing accounts more frequently than intended each time the mapping account reloads (and the batches get recreated)?

Reply:
this is not a big deal given that the mapping accounts reload very infrequently. maybe leave a // todo comment if this is the case.

Reply:
Nice catch

Reply:
This was an easy fix, just moved the mutex constructor to before the loop so that it and its state is reused. Async runtimes _must_ properly deallocate variables in scope of an aborted task (to avoid counter-intuitive mem leaks), so it should work well now and in the future